server/apps/publish/archive_publish.py

```diff
@@ -282,51 +303,94 @@ def publish(self, doc, updates, target_output_channels=None):
                 updates['source'] = desk['source'] if desk and desk.get('source', '') \
                     else DEFAULT_SOURCE_VALUE_FOR_MANUAL_ARTICLES
 
+            updates['pubstatus'] = PUB_STATUS.CANCELED if self.publish_type == 'killed' else PUB_STATUS.USABLE
+            updated.update(updates)
+
+        # Step 2(a)
+        if self.published_state == 'killed':
+            req = ParsedRequest()
+            req.sort = '[("completed_at", 1)]'
+            queued_items = get_resource_service('publish_queue').get(
+                req=req, lookup={'item_id': updated[config.ID_FIELD]})
+
+            if queued_items.count():
+                queued_items = list(queued_items)
+
+                # Step 2(a)(i)
+                subscribers = list(get_resource_service('subscribers').get(req=None, lookup=None))
+                recipients = [s['email'] for s in subscribers]
+                send_article_killed_email(doc, recipients, queued_items[0].get('completed_at'))
+
+                # Step 2(a)(ii)
+                no_formatters, queued = self.queue_transmission(updated, subscribers, None)
+        elif self.published_state == 'corrected':  # Step 2(b)
+            query = {'$and': [{'item_id': updated[config.ID_FIELD]}, {'publishing_action': 'published'}]}
+            queued_items = get_resource_service('publish_queue').get(req=None, lookup=query)
+
+            if queued_items.count():
+                # Step 2(b)(i)
+                queued_items = list(queued_items)
+                subscriber_ids = {queued_item['subscriber_id'] for queued_item in queued_items}
+
+                query = {'$and': [{config.ID_FIELD: {'$in': list(subscriber_ids)}}]}
+                subscribers = list(get_resource_service('subscribers').get(req=None, lookup=query))
+                no_formatters, queued = self.queue_transmission(updated, subscribers)
+
+                # Step 2(b)(ii)
+                active_subscribers = list(get_resource_service('subscribers').get(req=None, lookup={'is_active': True}))
```

Comment:
Probably redundant to convert to a list here? It's enough if `active_subscribers` can be iterated over (as it happens on the next line).